### PR TITLE
Support customizing the training loop using AllReduce.

### PR DESCRIPTION
**Pull request by workingloong**

### CODE REVIEW

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+the training loop when users use AllReduce distribution strategy.
+
+## Motivation
+
+Now, users need to define the forward computation, loss function,
+optimizer and dataset function in ElasticDL. ElasticDL provides
+the training loop with those definitions. Now ElasticDL only support
+Keras API to define the forward computation. It maybe not flexible for
+users to define complex models in CV or NLP. Sometimes, users need to
+customize the training loop to control the model iteration.
```
Comment (QiJune) [inline docs/designs/custom_allreduce_training_loop_support.md:13]: Could you please give an example of the demand for customizing the training loop?

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+
+If ElasticDL support customizing the training loop, users can use different
+deep learning library (e.g. TensorFlow, Pytorch) to define their training loop.
+ElasticDL only need to provide dynamic data partitioning for dataset and
+elastic AllReduce to merge gradients across workers.
+
+## The Training Loop of AllReduce
+
+The training loop of AllReduce contains two components:
```
Comment (QiJune) [inline docs/designs/custom_allreduce_training_loop_support.md]: two components --> two steps?

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+    inputs = tf.keras.Input(shape=(28, 28), name="image")
+    outputs = Conv(10)(inputs)
+
+    model = tf.keras.Model(inputs=inputs, outputs=outputs, name="mnist_model")
+
+    optimizer = tf.optimizers.SGD(lr)
+
+    # Set object to broadcast
+    ElasticBroadcastObject.set_model(model)
+    ElasticBroadcastObject.set_optimizer(optimizer)
```
Comment (QiJune) [inline docs/designs/custom_allreduce_training_loop_support.md]: Could you give a detailed explanation of `ElasticBroadcastObject`?

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
+# Support Customizing the Training Loop Using AllReduce
+
+This document describes the design for supporting customizing
+the training loop when users use AllReduce distribution strategy.
+
+## Motivation
+
+Now, users need to define the forward computation, loss function,
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: Two `now` within 3 sentences is not good.

`support` -> `supports`

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+We can develop a decorator to wrap the gradient combination process
+using AllReduce. The decorator can support fault-tolerance and
+elastic training. If AllReduce fails, the decorator will query the
+master for a new AllReduce ring and retry to combine gradients
+across alive workers. And the decorator also queries the master
+for a new ring periodically in case that the master add new workers.
+
+```python
+@elastic_allreduce
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: This code segment is confusing. If it is an example of how users to use elastic training with customized training loop,  the method definition should be consistent with the codes in the next section.

File: docs/designs/custom_allreduce_training_loop_support.md (hunk 1 above)
Comment (workingloong) [reply docs/designs/custom_allreduce_training_loop_support.md:13]: Given a pre-trained network, users might only want to optimize a new set of output weights.

File: docs/designs/custom_allreduce_training_loop_support.md (hunk 5 above)
Comment (workingloong) [reply docs/designs/custom_allreduce_training_loop_support.md]: I should delete the code segment.

File: docs/designs/custom_allreduce_training_loop_support.md (hunk 3 above)
Comment (workingloong) [reply docs/designs/custom_allreduce_training_loop_support.md]: The `ElasticBroadcastObject` is to contain objects which are needed to broadcast. I have deleted it and move it into the `ElasticAllReduceManager` in the design.

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 6]
```diff
@@ -0,0 +1,404 @@
+# Support Customizing the Training Loop Using AllReduce
+
+This document describes the design for supporting customizing
+the training loop when users use the AllReduce distribution strategy.
+
+## Motivation
+
+Users need to define the forward computation, loss function,
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: Users need to define a forward computation, a loss function, an optimizer, and a dataset function in ElasticDL.

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+optimizer, and dataset function in ElasticDL. ElasticDL provides
+the training loop with those definitions. ElasticDL only supports
+Keras API to define the forward computation. It maybe not flexible for
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: This is not very flexible for users to define some complex models in CV or NLP. Also, users may want to ....

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+users to define complex models in CV or NLP. Sometimes, users need to
+customize the training loop to control the model iteration. For example,
+For example, given a pre-trained network, users might only want to
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md:14]: Two `For example`

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+optimize a new set of output weights.
+
+If ElasticDL supports customizing the training loop, users can use different
+deep-learning library (e.g., TensorFlow, Pytorch) to define their training
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: library -> libraries

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+loop. ElasticDL only need to provide dynamic data partitioning for
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: need -> needs

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+dataset and elastic AllReduce to merge gradients across workers.
+
+## The Training Loop of ElasticDL
+
+The training loop of ElasticDL on the worker contains two steps:
+
+- Get tasks from the master to create dataset with data shards in tasks.
+- Execute training loop to calculate gradients and update the model
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md:27]: using the dataset. -> using gradients.

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+using the dataset.
+
+![Training Loop](../images/training_loop.jpg)
+
+From the figure, elastic training of ElasticDL focus on data partitioning
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: This part is not very clear. We can rewrite it:

In order to support customized training loop with ElasticDL AllReduce, ElasticDL provides users with a dataset with dynamic partition support, and an API `elastic_run` for elastic training. 

Users need to implement their own training logic for a training data batch in a function `train_one_batch`:

```Python
def train_one_batch(batch_data):
    loss = forward(batch_data)
    gradients = backward(loss)
    gradients = allreduce(gradients)
    update_model(gradients)
```

Then users can write the training loop using `train_one_batch`, the dataset and `elastic_run` API  provided by ElasticDL.

```python
def train(dataset, elastic_run):
    """
    Arguments:
        dataset: tf.data.Dataset. ElasticDL creates a dataset according to
            tasks from the master
        elastic_manager: A manager to execute elastic training.
    """
    for batch_data in dataset:
        elastic_train = elastic_run(train_one_batch)
        elastic_train(batch_data)
```

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+            mon_sess.run(train_op)
+```
+
+</td></tr>
+</table>
+
+### TensorFlow 2.x
+
+<table>
+<tr valign="top"><td> ElasticDL </td><td> Horovod </td></tr>
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md:212]: Codes order is wrong. ElasticDL code on the right now.

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 14]
```diff
@@ -0,0 +1,402 @@
...
+        ...
+```
+
+## Examples of Different Libraries Using ElasticDL AllReduce
+
+Users only need to make the following changes to their training loop:
+
+- Use the decorator of elastic training to wrap the code of gradient
+combination using Horovod.
+- Set objects to broadcast like the model and optimizer. Because an alive
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md:117]: How to `set_broadcast_object`? Any example?

Comment (workingloong) [reply docs/designs/custom_allreduce_training_loop_support.md:117]: I have listed the example in the following code snippets like:

```python
# Set object to broadcast
elastic_manager.set_model(mnist_model)
elastic_manager.set_optimizer(opt)
```

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 15]
```diff
@@ -0,0 +1,402 @@
...
+This document describes the design for supporting customizing
+the training loop when users use the AllReduce distribution strategy.
+
+## Motivation
+
+Users need to define a forward computation, a loss function, an optimizer,
+and a dataset function in ElasticDL. ElasticDL provides
+the training loop with those definitions. ElasticDL only supports
+Keras API to define the forward computation. This is not very flexible for
+users to define some complex models in CV or NLP. Also, users may wan to
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: wan to -> want to

File: docs/designs/custom_allreduce_training_loop_support.md [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+
+```python
+tape = hvd.DistributedGradientTape(tape)
+grads = tape.gradient(loss, model.trainable_variables)
+optimizer.apply_gradients(
+    zip(grads, model.trainable_variables)
+)
+```
+
+We can develop a decorator to wrap the gradient combination process
```
Comment (skydoorkai) [inline docs/designs/custom_allreduce_training_loop_support.md]: Thus, we can implement `elastic_run` as a decorator to wrap the gradient combination process using AllReduce.

**Review by skydoorkai — APPROVED** [review]
